Description: [PATCH] packagekit: Update package installed check to use new manual
 and auto statuses

The apt backend replaced the "installed" status with "auto" or "manual" in:
https://github.com/PackageKit/PackageKit/commit/cfd297aeb51bde67269d74c5a762634314e918b6

Without this change using when using PackageKit >= 1.2.5 .deb packages are not able to be uninstalled.

Fixes https://gitlab.gnome.org/GNOME/gnome-software/-/issues/1720
Author: Robert Ancell <robert.ancell@canonical.com>
Origin: other>, https://gitlab.gnome.org/Community/Ubuntu/gnome-software/-/commit/fe4703a989e1fd36b4afdf1116b7088c436f2ee4
Bug: https://gitlab.gnome.org/GNOME/gnome-software/-/issues/1720
Bug-Launchpad: https://bugs.launchpad.net/snap-store-desktop/+bug/1969303
Reviewed-by: David Mohammed <fossfreedom@ubuntu.com>
Last-Update: 2022-04-28
---
 plugins/packagekit/gs-plugin-packagekit.c | 27 ++++++++++++++++++++---
 1 file changed, 24 insertions(+), 3 deletions(-)

diff --git a/plugins/packagekit/gs-plugin-packagekit.c b/plugins/packagekit/gs-plugin-packagekit.c
index fba22ee3b..a8affa376 100644
--- a/plugins/packagekit/gs-plugin-packagekit.c
+++ b/plugins/packagekit/gs-plugin-packagekit.c
@@ -181,6 +181,27 @@ gs_plugin_destroy (GsPlugin *plugin)
 	g_object_unref (priv->task_upgrade);
 }
 
+static gboolean
+package_is_installed (const gchar *package_id)
+{
+	g_auto(GStrv) split = NULL;
+	const gchar *data;
+
+	split = pk_package_id_split (package_id);
+	if (split == NULL) {
+		return FALSE;
+	}
+
+	data = split[PK_PACKAGE_ID_DATA];
+	if (g_str_has_prefix (data, "installed") ||
+            g_str_has_prefix (data, "manual:") ||
+            g_str_has_prefix (data, "auto:")) {
+		return TRUE;
+	}
+
+	return FALSE;
+}
+
 static gboolean
 gs_plugin_add_sources_related (GsPlugin *plugin,
 			       GHashTable *hash,
@@ -463,7 +484,7 @@ gs_plugin_app_install (GsPlugin *plugin,
 		array_package_ids = g_ptr_array_new_with_free_func (g_free);
 		for (i = 0; i < source_ids->len; i++) {
 			package_id = g_ptr_array_index (source_ids, i);
-			if (g_strstr_len (package_id, -1, ";installed") != NULL)
+			if (package_is_installed (package_id))
 				continue;
 			g_ptr_array_add (array_package_ids, g_strdup (package_id));
 		}
@@ -478,7 +499,7 @@ gs_plugin_app_install (GsPlugin *plugin,
 			source_ids = gs_app_get_source_ids (addon);
 			for (j = 0; j < source_ids->len; j++) {
 				package_id = g_ptr_array_index (source_ids, j);
-				if (g_strstr_len (package_id, -1, ";installed") != NULL)
+				if (package_is_installed (package_id))
 					continue;
 				g_ptr_array_add (array_package_ids, g_strdup (package_id));
 			}
@@ -614,7 +635,7 @@ gs_plugin_app_remove (GsPlugin *plugin,
 	package_ids = g_new0 (gchar *, source_ids->len + 1);
 	for (i = 0; i < source_ids->len; i++) {
 		package_id = g_ptr_array_index (source_ids, i);
-		if (g_strstr_len (package_id, -1, ";installed") == NULL)
+		if (!package_is_installed (package_id))
 			continue;
 		package_ids[cnt++] = g_strdup (package_id);
 	}
-- 
2.34.1

